glean-core/ios/Glean/Metrics/StringMetric.swift

```diff
@@ -0,0 +1,96 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+/// This implements the developer facing API for recording string metrics.
+///
+/// Instances of this class type are automatically generated by the parsers at build time,
+/// allowing developers to record values that were previously registered in the metrics.yaml file.
+///
+/// The string API only exposes the `StringMetricType.set` method, which takes care of validating the input
+/// data and making sure that limits are enforced.
+public class StringMetricType {
+    let handle: UInt64
+    let disabled: Bool
+    let sendInPings: [String]
+
+    /// The public constructor used by automatically generated metrics.
+    public init(category: String, name: String, sendInPings: [String], lifetime: Lifetime, disabled: Bool) {
+        self.disabled = disabled
+        self.sendInPings = sendInPings
+        self.handle = withArrayOfCStrings(sendInPings) { pingArray in
+            glean_new_string_metric(
+                category,
+                name,
+                pingArray,
+                Int32(sendInPings.count),
+                lifetime.rawValue,
+                disabled ? 1 : 0
+            )
+        }
+    }
+
+    /// Destroy this metric.
+    deinit {
+        if self.handle != 0 {
+            glean_destroy_string_metric(self.handle)
+        }
+    }
+
+    /// Set a string value.
+    ///
+    /// - paramaters:
+    ///     * value: This is a user defined string value. If the length of the string exceeds
+    ///              the maximum length, it will be truncated.
+    public func set(_ value: String) {
+        guard !self.disabled else { return }
+
+        _ = Dispatchers.shared.launch {
+            self.setSync(value)
+        }
+    }
+
+    /// Internal only, synchronous API for setting a string value.
+    func setSync(_ value: String) {
+        glean_string_set(Glean.shared.handle, self.handle, value)
+    }
+
+    /// Tests whether a value is stored for the metric for testing purposes only. This function will
+    /// attempt to await the last task (if any) writing to the the metric's storage engine before
+    /// returning a value.
+    ///
+    /// - parameters:
+    ///     * pingName: represents the name of the ping to retrieve the metric for.
+    ///                 Defaults to the first value in `sendInPings`.
+    /// - returns: true if metric value exists, otherwise false
+    func testHasValue(_ pingName: String? = nil) -> Bool {
+        Dispatchers.shared.assertInTestingMode()
+
+        let pingName = pingName ?? self.sendInPings[0]
+        return glean_string_test_has_value(Glean.shared.handle, self.handle, pingName) != 0
+    }
+
+    /// Returns the stored value for testing purposes only. This function will attempt to await the
+    /// last task (if any) writing to the the metric's storage engine before returning a value.
+    ///
+    /// Throws a `String` exception if no value is stored
```

Comment:
wow, what's a `String` exception!?

Reply:
Just a String as an exception. On Kotlin we throw a NullPointerException I think, let me check if we can throw something similar

Reply:
No no, I think it's fine :) I was merely asking for curiosity. As a separate thing, did you check if docs show up in XCode when doing autocompletion?

Reply:
![auto completion on a metric](https://tmp.fnordig.de/scr/c8c9cdbeb8.png)

The first few items are internal and won't pop up when used from the outside (this is in the Glean tests, so it can see everything)